### PR TITLE
feat: Add and Delete project

### PR DESCRIPTION
**Pull request by VinayAdaki**
This PR will add new display Projects feature and will provide option to add new project.

image

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -1,18 +1,45 @@
...
+        {
+          title: 'Social App',
+          category: 'Mobile Development'
+        },
+        {
+          title: 'Ecommarce shopping Cart',
+          category: 'Web Development'
+        }
+      ]
+    });
```
Comment (KavitaLatad) [inline src/App.js]: You can achieve this setup in `constructor` as well.

File: src/App.js [hunk 2]
```diff
@@ -1,18 +1,45 @@
...
+  }
+  handleAddProject(project){
+    let projects = this.state.projects;
+    projects.push(project);
+    this.setState({projects:projects});
```
Comment (KavitaLatad) [inline src/App.js]: Can you try using shorthand code style here,

```
this.setState({
   projects
})
```

File: src/Components/AddProject.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import React, { Component } from 'react';
+
+class AddProject extends Component {
+  constructor(){
+      super();
+      this.state = {
+        newProject: {}
+      }
```
Comment (KavitaLatad) [inline src/Components/AddProject.js]: Can you please elaborate need for this `state `here

File: src/Components/AddProject.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        }, function(){
+          //console.log(this.state);
+          this.props.addProject(this.state.newProject);
+        })
+      }
+      e.preventDefault();
+  }
+
+  render() {
+    let categoryOptions = this.props.categories.map(category => {
```
Comment (KavitaLatad) [inline src/Components/AddProject.js]: you can use `const` here instead of `let`

File: src/Components/Projects.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  render() {
+    let projectItems;
+    if(this.props.projects){
+        projectItems = this.props.projects.map(project => {
+          //console.log(project);
+          return (
+            <ProjectItem key={project.title} project={project} />
+          );
+        });
+    }
```
Comment (KavitaLatad) [inline src/Components/Projects.js]: We can also do this in `return` block below.

**Comment by KavitaLatad** [comment]
@VinayAdaki Good job !!! Just a side note here, I can't see `props `documented in the components, which is little bit confusing. I suggest its a good practice to document the props.

File: src/App.js (hunk 1 above)
Comment (VinayAdaki) [reply src/App.js]: In the earlier session, this was in the Constructor only, but he wanted to show lifecycle methods here.
I will update this once after the complete session.

File: src/Components/Projects.js (hunk 5 above)
Comment (VinayAdaki) [reply src/Components/Projects.js]: Good to know. But what I am thinking Return should be clean.
will it make any difference here in terms of performance or clean code practice?

File: src/Components/AddProject.js (hunk 4 above)
Comment (VinayAdaki) [reply src/Components/AddProject.js]: I will try it out. just come to know the difference again.
https://wesbos.com/let-vs-const/

**Comment by VinayAdaki** [comment]
I will try to add once I get some more understanding of it, But for sure before merging or closing os this PR.

File: src/Components/AddProject.js (hunk 3 above)
Comment (VinayAdaki) [reply src/Components/AddProject.js]: I want to push new New project object to state so I can retrieve and pass it back to call back function.

**Comment by VinayAdaki** [comment]
[How React Works](https://www.youtube.com/watch?v=mLMfx8BEt8g)
ReactLifecycle Methods 
- [Reference1 with Diagram](https://hackernoon.com/reactjs-component-lifecycle-methods-a-deep-dive-38275d9d13c0)
- [Reference2](https://engineering.musefind.com/react-lifecycle-methods-how-and-when-to-use-them-2111a1b692b1)

File: src/App.js [hunk 6]
```diff
@@ -1,18 +1,56 @@
...
+  }
+  handleAddProject(project){
+      let projects = this.state.projects;
+      projects.push(project);
+      this.setState({projects});
+  }
+
+  handleDeleteProject(id){
+      let projects = this.state.projects;
+      let index = projects.findIndex(project => project.id === id);
```
Comment (KavitaLatad) [inline src/App.js]: this should be a `const`

File: src/Components/AddProject.js (hunk 3 above)
Comment (KavitaLatad) [reply src/Components/AddProject.js]: You can achieve this using a simple object variable.

File: src/Components/AddProject.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+      return <option key={category} value={category}> {category} </option>
+    });
+    return (
+      <div>
+        <h3>Add Project</h3>
+        <form onSubmit={this.handleSubmit.bind(this)}>
+          <div>
+            <label>Title</label>
+            <input type="text" ref="title" />
```
Comment (KavitaLatad) [inline src/Components/AddProject.js]: Its is not good practice to use `ref`. [Read More...](https://www.robinwieruch.de/react-ref-attribute-dom-node/#when) Use `id` or `class` instead

File: src/Components/AddProject.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+          <br/>
+        </form>
+      </div>
+    );
+  }
+}
+
+AddProject.propTypes = {
+  categories: PropTypes.array,
+  addProject: PropTypes.func
```
Comment (KavitaLatad) [inline src/Components/AddProject.js]: Are theses props `required`

File: src/Components/Projects.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
+import React, { Component } from 'react';
+import ProjectItem from './ProjectItem';
+import PropTypes from 'prop-types'
+class Projects extends Component {
+  onDelete(id){
```
Comment (KavitaLatad) [inline src/Components/Projects.js]: Do you really need this

File: src/Components/Projects.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+      this.props.onDelete(id);
+  }
+
+  render() {
+    let projectItems;
+    if(this.props.projects){
+        projectItems = this.props.projects.map(project => {
+          //console.log(project);
```
Comment (KavitaLatad) [inline src/Components/Projects.js]: Please delete this comment

File: src/Components/Projects.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+          return (
+            <ProjectItem onDelete={this.onDelete.bind(this)} key={project.title} project={project} />
```
Comment (VinayAdaki) [inline src/Components/Projects.js]: Try deleting `this.onDelete.bind(this)` and use `this.props.onDelete`

File: src/Components/Projects.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+        <h3>Latest Projects</h3>
+        {projectItems}
+      </div>
+    );
+  }
+}
+
+Projects.propTypes = {
+  projects: PropTypes.array,
+  onDelete: PropTypes.func
```
Comment (VinayAdaki) [inline src/Components/Projects.js]: try using isRequired on PropTypes.x

File: src/Components/Projects.js (hunk 10 above)
Comment (VinayAdaki) [inline src/Components/Projects.js]: I hate this

File: src/Components/AddProject.js (hunk 7 above)
Comment (VinayAdaki) [inline src/Components/AddProject.js]: instead of `ref` use `class or id`

File: src/Components/AddProject.js [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+      if(this.refs.title.value === ''){
+          alert('Title is required');
+      } else {
+          this.setState({newProject: {
+            id: uuid.v4(),
+            title: this.refs.title.value,
+            category: this.refs.category.value
+          }
+        }, function(){
+          //console.log(this.state);
```
Comment (VinayAdaki) [inline src/Components/AddProject.js]: :-1:

File: src/Components/AddProject.js (hunk 3 above)
Comment (VinayAdaki) [reply src/Components/AddProject.js]: Please provide some reference.

**Review by KavitaLatad — APPROVED** [review]

**Pull request merged**